modules/sitemaps/sitemaps-provider-decorator.php

```diff
@@ -59,29 +79,113 @@ public function get_url_list( $page_num, $object_subtype = '' ) {
+	}
+
+	/**
+	 * Get active languages for the sitemap.
+	 *
+	 * @since 2.8
+	 */
+	protected function get_active_languages() {
+		$languages = $this->model->get_languages_list();
+		if ( wp_list_filter( $languages, array( 'active' => false ) ) ) {
+			return wp_list_pluck( wp_list_filter( $languages, array( 'active' => false ), 'NOT' ), 'slug' );
 		}
+		return wp_list_pluck( $languages, 'slug' );
+	}
 
-		return $this->provider->get_max_num_pages( $object_subtype );
+	/**
+	 * Filters the query arguments to add the language.
+	 *
+	 * @since 2.8
+	 *
+	 * @param array $args Sitemap provider WP_Query or WP_Term_Query arguments.
+	 * @return array
+	 */
+	public function query_args( $args ) {
+		if ( isset( $this->filter_lang ) ) {
+			$args['lang'] = $this->filter_lang;
+		}
+		return $args;
+	}
+
+	/**
+	 * Gets data for a given sitemap type.
+	 *
+	 * @since 2.8
+	 *
+	 * @param string $object_subtype_name Object subtype name if any.
+	 * @param string $lang                Optionnal language name.
+	 * @return array
+	 */
+	protected function get_sitemap_data( $object_subtype_name, $lang = '' ) {
+		$object_subtype_name = (string) $object_subtype_name;
+
+		if ( ! empty( $lang ) ) {
+			$this->filter_lang = $lang;
+		}
+
+		$return = array(
+			'name'  => implode( '-', array_filter( array( $object_subtype_name, $lang ) ) ),
+			'pages' => $this->get_max_num_pages( $object_subtype_name ),
+		);
+
+		unset( $this->filter_lang );
+		return $return;
+	}
+
+	/**
+	 * Gets data about each sitemap type.
+	 *
+	 * @since 2.8
+	 *
+	 * @return array[] Array of sitemap types including object subtype name and number of pages.
+	 */
+	public function get_sitemap_type_data() {
+		$sitemap_data = array();
+
+		add_filter( 'wp_sitemaps_posts_query_args', array( $this, 'query_args' ) );
+		add_filter( 'wp_sitemaps_taxonomies_query_args', array( $this, 'query_args' ) );
+
+		$object_subtypes = $this->get_object_subtypes();
+
+		if ( empty( $object_subtypes ) ) {
+			foreach ( $this->get_active_languages() as $language ) {
+				$sitemap_data[] = $this->get_sitemap_data( '', $language );
+			}
+		}
+
+		foreach ( array_keys( $object_subtypes ) as $object_subtype_name ) {
+			switch ( $this->provider->name ) {
+				case 'posts':
+					if ( $this->model->is_translated_post_type( $object_subtype_name ) ) {
+						foreach ( $this->get_active_languages() as $language ) {
+							$sitemap_data[] = $this->get_sitemap_data( $object_subtype_name, $language );
+						}
+					} else {
+						$sitemap_data[] = $this->get_sitemap_data( $object_subtype_name );
+					}
+					break;
+				case 'taxonomies':
+					if ( $this->model->is_translated_taxonomy( $object_subtype_name ) ) {
+						foreach ( $this->get_active_languages() as $language ) {
+							$sitemap_data[] = $this->get_sitemap_data( $object_subtype_name, $language );
+						}
+					} else {
+						$sitemap_data[] = $this->get_sitemap_data( $object_subtype_name );
+					}
+					break;
+			}
```

Comment:
The only change between these two block is the method call in the condition. You can remove duplication by making it a variable, computed before the `foreach` loop.